src/poetry/utils/password_manager.py

```diff
@@ -195,18 +195,16 @@ def delete_pypi_token(self, repo_name: str) -> None:
     def get_http_auth(self, repo_name: str) -> dict[str, str | None] | None:
         username = self._config.get(f"http-basic.{repo_name}.username")
         password = self._config.get(f"http-basic.{repo_name}.password")
-        if not username and not password:
+
+        if not username:
             return None
 
-        if not password:
-            if self.use_keyring:
-                password = self.keyring.get_password(repo_name, username)
-            else:
-                return None
+        if not password and self.use_keyring:
+            password = self.keyring.get_password(repo_name, username)
 
         return {
             "username": username,
-            "password": password,
+            "password": password or "",
```

Comment:
the type annotation on the return value for this function can be tightened up now I think - the values in this dictionary are guaranteed strings

Reply:
actually perhaps it should have been an `HTTPAuthCredential` all along, though perhaps we are beginning to get into feature creep for this pull request

Reply:
While it is a bit of a feature creep, I reckon it might be a useful one. I'll try going down the rabbit hole to see how far I can go without it becoming a much bigger change.

Reply:
Turns out sending empty string when password is None is probably a bad thing. But fixed the method to return `HTTPAuthCredential` now. Further tightening might require a slightly more involved PR.

Reply:
Yeah.  I would think the `HTTPAuthCredential` should guarantee a non-empty - or at least, not-`None` - username.  But I can definitely see that being a change that begins to touch more than you might want to right now!

Reply:
Yeah will look into the side-effects of that a bit later. For now this should suffice.